core/src/main/java/hudson/slaves/JNLPLauncher.java

```diff
@@ -249,4 +255,23 @@ public FormValidation doCheckWebSocket(@QueryParameter boolean webSocket, @Query
     public boolean isJavaWebStartSupported() {
         return JavaUtils.isRunningWithJava8OrBelow();
     }
+
+    /**
+     * Overrides the url that inbound TCP agents should connect to
+     * as advertised in the agent.jnlp file. If not set, the default
+     * behavior is unchanged and returns the root URL.
+     *
+     * This enables using a private address for inbound tcp agents,
+     * separate from Jenkins root URL.
+     *
+     * @see @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-63222">JENKINS-63222</a>
+     */
+    @Restricted(NoExternalUse.class)
+    public static String getInboundAgentUrl() {
+        String url = System.getProperty(CUSTOM_INBOUND_URL_PROPERTY);
```

Comment:
Please use https://javadoc.jenkins.io/jenkins/util/SystemProperties.html so that other configuration providers could be used as well